### ISSUE TITLE
Closures Handling

### ISSUE THREAD

**Issue by na2axl** [gircore/gir.core#119]
As discussed in #103, this issue is to find a solution about closures handling in Gir.Core libraries.
This issue can be related to #109 

---

_**@firox263:** Borrowing this to add a TODO_

**TODO 6/05/21:**
 - [ ] Reimplement Native->Managed marshalling (so we can call the user's provided delegate)
 - [ ] Properly exclude certain parameters like `userData`
 - [ ] Write unit/integration tests

**Completed:**
 - [X] Prototype Notified
 - [X] Prototype Async
 - [x] Prototype Default/Call
 - [x] Generator should use Native-suffix in native function calls
 - [x] Generator needs a way to marshal any arbitrary type from Native->Managed (this may require the generator rewrite to achieve?)
 - [x] Come to a decision whether DestroyNotify is useful in the end user API
 - [x] Implement Notify in Generator
 - [x] Implement Async in Generator
 - [x] Implement Call in Generator
 - [x] Fix DBusConnection and a few existing uses
 - [x] Memory Leak Testing

**Comment by firox263** [comment]
As part of this, we should also tackle delegate/callback signatures. Currently we generate delegates with IntPtr for most parameters, but obviously we wouldn't want users to have to marshal themselves. We should modify the generator to generate a managed delegate and associated helper class for each callback which handles marshalling automatically (similar to how we populate TSignalArgs).

If we implement this using closures we can probably kill two birds with one stone here, and also reduce the amount of boilerplate we have to write in methods.

**Comment by firox263** [comment]
Some ideas on how to handle delegate resources (to avoid the GC):

There is a scope parameter in GIR metadata for parameter.

See https://developer.gnome.org/gi/unstable/gi-GIArgInfo.html#GIScopeType

It has four possible values:
 - Invalid (not a callback)
 - Call (only used during the function - can free in-func)
 - Async (only ever called once - free after use)
 - Notified (DestroyNotify to free the callback resources)

For example, GtkTickCallback:

```xml
<parameter name="callback" transfer-ownership="none" scope="notified" closure="1" destroy="2">
    <doc xml:space="preserve" filename="gtk/gtkwidget.c" line="5301">function to call for updating animations</doc>
    <type name="TickCallback" c:type="GtkTickCallback"/>
</parameter>
```

From https://gi.readthedocs.io/en/latest/annotations/giannotations.html

 - call (default) - Only valid for the duration of the call. Can be called multiple times during the call.
 - async - Only valid for the duration of the first callback invocation. Can only be called once.
 - notified - valid until the GDestroyNotify argument is called. Can be called multiple times before the GDestroyNotify is called.

**Comment by firox263** [comment]
Pseudocode example:

```cs
// Call:

void SomeFunctionCall(tick_callback)
{
    // Allocate
    var thing = new DefaultCallback<TickCallback>(callback);

    // Native Call
    Native.some_function(thing);

    // Free
    Free(thing);
}
```

```cs
// Async:

void SomeFunctionAsync(some_async_func)
{
    var thing = new AsyncCallback<SomeAsyncFunc>(some_async_func);
    Native.some_function(thing.GetCallback());
    Free(thing);
}

// This is hardcoded at generation time
class AsyncCallback<T>
{
    public AsyncCallback<T>(T callback)
    {
        GCH.Root(callback)
    }

    InvokeManagedFunction()
    {
        callback();
        // free resources here because it'll never be called again
    }
}
```

**Comment by firox263** [comment]
Big update as a result of some research/testing:
 * We will need a Default/Async/Notified handler for _each_ callback, as the scope is a property of the parameter, not the callback (i.e. the same callback can be used with any type of scope).
 * To avoid runtime overhead, it's best if we just generate the three different handlers for each type of callback.

This is a working example of a Notified-scope handler for Gst.BusFunc (hand written for now, the generator needs changes to support this):

```cs
// BusFunc.Generated.cs
namespace Gst
{
    public delegate bool BusFunc(Bus bus, Message message);
    
    public delegate bool BusFuncNative(IntPtr bus, IntPtr message, IntPtr user_data);

    public class BusFuncHandlerNotified
    {
        public event Action OnDestroyNotify;
        public BusFuncNative NativeCallback;

        private BusFunc managedCallback;
        private DestroyNotify DestroyNotify;
        private GCHandle gch;

        public BusFuncHandlerNotified(BusFunc managed)
        {
            DestroyNotify = DestroyCallback;
            NativeCallback = NativeCallbackMarshaller;
            managedCallback = managed;
            gch = GCHandle.Alloc(this);
        }

        private bool NativeCallbackMarshaller(IntPtr bus, IntPtr message, IntPtr userData)
        {
            Bus busParameter = Object.WrapHandle<Bus>(bus);
            Message messageParameter = Marshal.PtrToStructure<Message>(message);
            return managedCallback(busParameter, messageParameter);
        }

        private void DestroyCallback(IntPtr userData)
        {
            OnDestroyNotify();

            // Allow for garbage collection
            gch.Free();
        }
    }
}
```
```cs
// Bus.cs
namespace Gst
{
    public partial class Bus
    {
         public uint AddWatchFull(int priority, BusFunc func)
            => AddWatchFull(priority, func, null);

        public uint AddWatchFull(int priority, BusFunc func, Action? destroyNotify)
        {
            var handler = new BusFuncHandlerNotified(func);

            if (destroyNotify != null)
                handler.OnDestroyNotify += destroyNotify;
            
            return Native.add_watch_full(Handle, priority, handler.NativeCallback, IntPtr.Zero, handler.DestroyNotify);
        }
    }
}
```

I haven't yet tested if resources are freed appropriately, but it appears to stop the GC problem for this specific function, and doesn't have any apparent regressions. Will prototype some more then look at integrating this into the generator.

I'm not sure the current generator design is capable of implementing this (particularly marshalling from arbitrary unmananged types -> managed types), so I might start working on #100 at the same time.

See the top for a TODO list.

**Comment by badcel** [comment]
I think the concept can work in general :+1:. I wonder if the watch should be removed if the bus object is disposed as `add_watch_full`  has the bus as an instance parameter?

**Comment by badcel** [comment]
I think the current generator should be able to at least generate the helper class quite easily as it would only be a custom scriban script for each delegates which generate a new file. The usage of the helper could be hand written for now. The implementation of the new generator will probably take some time (probably month) so I would prefer the simple solution for now.

**Comment by badcel** [comment]
Another idea, we should verify if we can leverage function pointers and replace the delegates with them. Perhaps this makes this problem a lot easier.

**Comment by badcel** [comment]
@firox263 I do not know how to work with `bus.AddWatchFull` but this should probably work from a technical point of view:

```cs
public unsafe partial class Bus
{
    private Func<Message, bool>? _watch;
    
    public uint AddWatchFull(int priority, Func<Message, bool> func)
    {
        _watch = func;
        
        [UnmanagedCallersOnly]
        static bool Callback(IntPtr obj, IntPtr message, IntPtr bla)
        {
            if (!GetObject(obj, out Bus bus))
                throw new Exception();

            if (bus._watch is null)
                throw new Exception();

            var m = Marshal.PtrToStructure<Message>(message);

            return bus._watch(m);
        }
        
        return Native.add_watch_full(Handle, priority, &Callback, IntPtr.Zero, null!);
    }
}
```

```cs
DllImport("Gst", EntryPoint = "gst_bus_add_watch_full")]
public static unsafe extern uint add_watch_full(IntPtr bus, int priority, delegate* unmanaged<IntPtr, IntPtr, IntPtr, bool> func, IntPtr user_data, GLib.DestroyNotify notify);
```

Can you provide some short sample what must be done to receive the message callbacks?

**Comment by firox263** [comment]
> I think the concept can work in general 👍. I wonder if the watch should be removed if the bus object is disposed as `add_watch_full` has the bus as an instance parameter?

I don't think we should, as when our Bus object is disposed, we might not be the only owner of the bus (as in, GStreamer may hold on to the bus and even return it to us later), So removing the watch could end up causing unexpected behaviour.

>  We probably need to change the currently existing code, too and make some kind of API review. Meaning I would like to have some kind of a partially working prototype which covers all major features we want to support before starting the rewrite.

Fully agree. I'm thinking of the generator rewrite as being "version 0.1" if you will, which contains the minimum set of functionality needed to wrap GObject.

> I think the current generator should be able to at least generate the helper class quite easily as it would only be a custom scriban script for each delegates which generate a new file. The usage of the helper could be hand written for now. The implementation of the new generator will probably take some time (probably month) so I would prefer the simple solution for now.

I'm not actually sure this is actually possible. In this part of the template:

```cs
private bool NativeCallbackMarshaller(IntPtr bus, IntPtr message, IntPtr userData)
{
    Bus busParameter = Object.WrapHandle<Bus>(bus);
    Message messageParameter = Marshal.PtrToStructure<Message>(message);
    return managedCallback(busParameter, messageParameter);
}
```

We need to, for any arbitrary type (e.g. `IntPtr message`), be able to know:
 * This should be marshalled to the managed `Message` type
 * Whether Message is a struct (`Marshal.PtrToStructure`), an object (`Object.WrapHandle`), or an enum (straightforward cast).

Since we don't have any rich type information in the generator, I'm not sure how feasible this is. It's similar to the interface prefixing problem, where we need to access and modify our knowledge of types at generation time.

> Can you provide some short sample what must be done to receive the message callbacks?

From Banshee (specifically `Banshee.GStreamerSharp/PlayerEngine.cs`):

```cs
public void Initialize()
{
    // We add a watch to our playbin
    playbin = ElementFactory.Make ("playbin", "the playbin");
    playbin.GetBus().AddWatchFull (Priority.Default, OnBusMessage);
}

// We declare a method which matches the callback signature
private bool OnBusMessage (Bus bus, Message msg)
{
    switch (msg.type) {
        case MessageType.Eos:
            StopIterating ();
            Close (false);
            OnEventChanged (PlayerEvent.EndOfStream);

            OnEos ();
            break;

        case MessageType.StateChanged:
            if (msg.Src == playbin) {
                State? old_state, new_state, pending_state;
                msg.ParseStateChanged (out old_state, out new_state, out pending_state);
                HandleStateChange (old_state, new_state, pending_state);
            }
            break;

        // ...
    }
}
```

> @firox263 I do not know how to work with bus.AddWatchFull but this should probably work from a technical point of view:

We need to be able to add any number of watches via the `AddWatchFull` function, so doing it like that wouldn't really work. There's also the problem of needing to free resources for only certain callbacks (as the user/runtime can remove our watch whenever it wants). Also, generally speaking, I'd like to avoid unsafe unless we really need it. If we find a performance hit at any point, we can always switch to it later.

**Comment by badcel** [comment]
> I don't think we should, as when our Bus object is disposed, we might not be the only owner of the bus (as in, GStreamer may hold on to the bus and even return it to us later), So removing the watch could end up causing unexpected behaviour.

Yes but wouldn't we use toggle ref to bind the lifetime of our object to the C object? So we would not get disposed if the user is not actively doing it. If the bus get's actively disposed by the user it means it should be destroyed, even on the C side.

As every GObject is disposable I think we should respect it's "expected" behavior. If the user disposes an object it is okay that unexpected behavior occurs. Because relying on a disposed object is in itself an error. 
I'm not 100% sure but if we do it like in your example it could probably happen that we never free the `BusFuncHandlerNotified ` object although it's lifetime is bound to the bus object on C side and on C# side. Meaning C side is freed, C# is garbage collected but the `BusFuncHandlerNotified ` object is still around because the manually added handle never got freed.

> Fully agree. I'm thinking of the generator rewrite as being "version 0.1" if you will, which contains the minimum set of functionality needed to wrap GObject.

This is fine for me :+1:. My alternative suggestion would be to add some helper function to the Script objects which allows to access the GNamespace object in scriban. Then the data should be available (for the currently processed gir file). It could be some helper method either like `GetClass` (search for class) `GetStruct` (search for struct) or s.th like `Get` (return class / struct whatever is found with this name) and `GetKind` which returns some enum for `class` / `struct`. I think my suggestion is easier but probably will fall short some day perhaps even while you code. 
So I think you should decide yourself which solution you prefer I just wanted to state an alternative.

In case you want to start with the rewrite I would like to request that you treat the "IntPtr wrapping code" code more like a "service". You already suggested that we can implement some kind of "staged" generator what I like.
Furthermore I want to introduce the concepts of "services". Meaning small pieces of code which do one think have as little dependencies as possible and could be exchanged in case of need. This should increase maintainablity as code is more seperated. The "IntPtr wrapping service" itself could use other smaller services to reach it's goal. If we structure the code from the beginning in this way there should not be the mess we have in the current generator where a lot of things happen in one class what makes it confusing. The hard part will be to separate the services in a meaningful manner.

>We need to be able to add any number of watches via the AddWatchFull function, so doing it like that wouldn't really work. 

For a general solution you are right, we could just use one dictionary per callback func. In this specific case there will never be more than one watch inside the dictionary as the documentations states:

> There can only be a single bus watch per bus, you must remove it before you can set a new one.

The dictionary would probably be cleared in case of the object / bus disposal. Avoiding a leaked object on C# side.

>  Also, generally speaking, I'd like to avoid unsafe unless we really need it. If we find a performance hit at any point, we can always switch to it later.

I have no fixed opinion on this. But I thought we already agreed to tend to use `unsafe` code for the native side (see #109) to avoid marshalling which should make it easier to generate the code and be faster. But we can decide which parts to be marshalled and which to be native in a case by case manner as the different situations should be isolated from each other.

**Comment by firox263** [comment]
> **badcel:** Yes but wouldn't we use toggle ref to bind the lifetime of our object to the C object? So we would not get disposed if the user is not actively doing it. If the bus get's actively disposed by the user it means it should be destroyed, even on the C side.

I don't think we should (for wrapper objects at least). GObjects tend to handle their own cleanup, so as long as we unref/clear where appropriate, there shouldn't be any problems.

We'd only use toggle refs for subclass objects as they contain important (managed) state which is not captured by the gobject. Wrapper types are simply a thin wrapper over an IntPtr handle, so we can destroy and recreate them as necessary. Generally speaking, toggle refs are not desirable and cause a lot of trouble (e.g. in GJS), so we'd only use them where necessary. From the docs for [`g_object_add_toggle_ref`](https://developer.gnome.org/gobject/stable/gobject-The-Base-Object-Type.html#g-object-add-toggle-ref):

> Multiple toggle references may be added to the same gobject, however if there are multiple toggle references to an object, none of them will ever be notified until all but one are removed. **For this reason, you should only ever use a toggle reference if there is important state in the proxy object.**

Where important state = managed class state, which is only present on subclasses, as wrappers just call into GObject for state.

Specifically in the case of GstBus, we shouldn't make any assumptions about cleanup because disposing of the bus (analogus to unreffing the bus) does not actually destroy it. From [gst_bus_add_watch_full](https://gstreamer.freedesktop.org/documentation/gstreamer/gstbus.html#gst_bus_add_watch_full):

> The bus watch will take its own reference to the bus, so it is safe to unref bus using gst_object_unref after setting the bus watch.

For the watch itself, it says:

> Return FALSE if the event source should be removed.

So in the case of gstreamer watches, the watch itself determines it should be removed by returning false, at which point GObject will then send the DestroyNotify event. So we don't need to worry about memory leaks upon disposing, as GStreamer keeps a reference to the bus anyway, so both the bus and our watch will still be alive. The bus is only destroyed when all references are released, including the one that we own (through the proxy object), and the one that GStreamer owns for the watch. When the user removes the source (returning false), this reference should be released.

>  I'm not 100% sure but if we do it like in your example it could probably happen that we never free the BusFuncHandlerNotified object although it's lifetime is bound to the bus object on C side and on C# side. Meaning C side is freed, C# is garbage collected but the BusFuncHandlerNotified object is still around because the manually added handle never got freed.

In my prototype, the manually added handle gets freed in destroy notify, which gets called whenever the async source is removed. 

I made a small test here in C ([firox263/destroy-notify.c](https://gist.github.com/firox263/441720c9a53520411d47d2ff43938393)) which shows that destroy notify is called by returning FALSE. So I don't think there's a chance we'd having a floating delegate handler, as the user must explicitly signal to stop iterating, and the library takes care to notify us when this happens.

---

I like your idea of services a lot 👍 I've been playing around with the generator a bit, and I might try and land some initial work alongside delegates, as I'm not sure if it's worth trying to hack a type dictionary into the existing generator (for marshalling).

> In this specific case there will never be more than one watch inside the dictionary 

Ah yes you're right, missed that. I wonder what happens if you try add another one? Maybe it just returns zero and silently fails.

> I have no fixed opinion on this. But I thought we already agreed to tend to use unsafe code for the native side (see #109) to avoid marshalling which should make it easier to generate the code and be faster? 

I was meaning more in the mean time while we still have lots of moving parts. Makes it easier to debug 😅. No problems with using more unsafe long term, as long as we can test it reliably.

**Comment by badcel** [comment]
>> For this reason, you should only ever use a toggle reference if there is important state in the proxy object.

>Where important state = managed class state, which is only present on subclasses, as wrappers just call into GObject for state.

You are right, according to the documentation this sounds like the right approach. I think there are multiple ways forward but we could just start with this one  and see how it goes. I opened a new issue for this see #185

-------------

Regarding the watch you are right i did not fully understand the API. The API is like guaranteeing that the bus is available as long as the destroy notify is not called.

**Comment by badcel** [comment]
@firox263 can we close this with #323?

**Comment by mjakeman** [comment]
> @firox263 can we close this with #323?

Yes, closures should be fully supported now. Anything remaining would fall under "general correctness", so I'd say we can go ahead and close this
